### PR TITLE
tcp_server example, read till the end

### PR DESCRIPTION
**Pull request by goriunov**
Read data till would block or closed connection with extendable vector.

### CODE REVIEW

File: examples/tcp_server.rs [hunk 1]
```diff
@@ -112,33 +112,42 @@ fn handle_connection_event(
...
                 }
+                Ok(_n) => received_data.extend_from_slice(&buf),
+                // Would block "errors" are the OS's way of saying that the
+                // connection is not actually ready to perform this I/O operation.
+                Err(ref err) if would_block(err) => break,
+                // Got interrupted (how rude!), we'll try again.
+                // NOTE: that this is not the proper way to this as we'll also redo
+                // the writable side above.
+                Err(ref err) if interrupted(err) => {
+                    return handle_connection_event(poll, connection, event)
```
Comment (Thomasdezeeuw) [inline examples/tcp_server.rs]: Can you change this to a `continue` (and remove the note above).

File: examples/tcp_server.rs [hunk 2]
```diff
@@ -112,33 +112,42 @@ fn handle_connection_event(
...
+        loop {
+            let mut buf = [0; 256];
+            match connection.read(&mut buf) {
+                Ok(0) => {
+                    // Reading 0 bytes means the other side has closed the
+                    // connection or is done writing, then so are we.
+                    connection_closed = true;
+                    break;
                 }
+                Ok(_n) => received_data.extend_from_slice(&buf),
```
Comment (Thomasdezeeuw) [inline examples/tcp_server.rs]: You need to use `&buf[..n]` because only `n` bytes are valid (the others are zero).

File: examples/tcp_server.rs [hunk 3]
```diff
@@ -112,33 +112,42 @@ fn handle_connection_event(
     }
 
     if event.is_readable() {
+        let mut connection_closed = false;
+        let mut received_data = Vec::with_capacity(4056);
```
Comment (Thomasdezeeuw) [inline examples/tcp_server.rs]: Can you change this to 4096 (4056 is an odd capacity).

File: examples/tcp_server.rs (hunk 1 above)
Comment (goriunov) [reply examples/tcp_server.rs]: Done :)

File: examples/tcp_server.rs (hunk 3 above)
Comment (goriunov) [reply examples/tcp_server.rs]: :man_facepalming:

File: examples/tcp_server.rs (hunk 2 above)
Comment (goriunov) [reply examples/tcp_server.rs]: Done

**Comment by Thomasdezeeuw** [comment]
Thanks.

**Pull request merged**